src/couch_views/src/couch_views_indexer.erl

```diff
@@ -171,20 +171,42 @@ map_docs(Mrst, Docs) ->
     % Run all the non deleted docs through the view engine and
     Mrst1 = start_query_server(Mrst),
     QServer = Mrst1#mrst.qserver,
-    MapFun = fun
-        (#{deleted := true} = Change) ->
-            Change#{results => []};
-        (#{deleted := false} = Change) ->
-            #{doc := Doc} = Change,
-            couch_stats:increment_counter([couchdb, mrview, map_doc]),
-            {ok, RawResults} = couch_query_servers:map_doc_raw(QServer, Doc),
-            JsonResults = couch_query_servers:raw_to_ejson(RawResults),
-            ListResults = lists:map(fun(ViewResults) ->
-                [list_to_tuple(Res) || Res <- ViewResults]
-            end, JsonResults),
-            Change#{results => ListResults}
-    end,
-    {Mrst1, lists:map(MapFun, Docs)}.
+
+    {Deleted0, NotDeleted0} = lists:partition(fun(Doc) ->
+        #{deleted := Deleted} = Doc,
+        Deleted
+    end, Docs),
+
+    Deleted1 = lists:map(fun(Doc) ->
+        Doc#{results => []}
+    end, Deleted0),
+
+    DocsToMap = lists:map(fun(Doc) ->
+        #{doc := DocRec} = Doc,
+        DocRec
+    end, NotDeleted0),
+
+    {ok, AllResults} = couch_eval:map_docs(QServer, DocsToMap),
+
+    % The expanded function head here is making an assertion
+    % that the results match the given doc
+    NotDeleted1 = lists:zipwith(fun(#{id := DocId} = Doc, {DocId, Results}) ->
+        Doc#{results => Results}
+    end, NotDeleted0, AllResults),
+
+    % I'm being a bit careful here resorting the docs
+    % in order of the changes feed. Theoretically this is
+    % unnecessary since we're inside a single transaction.
+    % However, I'm concerned if we ever split this up
+    % into multiple transactions that this detail might
+    % be important but forgotten.
+    MappedDocs = lists:sort(fun(A, B) ->
```

Comment:
I guess this is something for later. But we doing a lot of looping here. Won't that slow down performance of the index build?

Reply:
That'll be something to check with tracing once that's an option I'd think.